Allows to combine the `-f` and `-l` flags in kubectl logs

m1kola opened the pull request:
**What this PR does / why we need it**:

This PR allows to combine the `-f` and `-l` flags in the `kubectl logs` by reading logs from multiple sources simultaneously. It is a part of what was requested in #52218 (without the part about `-c`).

Example:

```
kubectl logs -l app=logging_test --all-containers -f
```

**Release note**:
```release-note
It is now possible to combine the `-f` and `-l` flags in `kubectl logs`
```

---

/sig cli
/kind feature

m1kola committed:
Allows to read from multiple logs simultaneously

Which makes possible to combile the `-f` and `-l` flags in kubectl logs

neolit123 commented:
/ok-to-test

m1kola commented:
/assign @liggitt

liggitt commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -293,13 +289,32 @@ func (o LogsOptions) RunLogs() error {
-			return err
-		}
+		go func(request *rest.Request) {
+			if err := o.ConsumeRequestFn(request, pipew); err != nil {
```
does this interleave log lines between sources? does it interleave sub-line?

liggitt replied:
is it unexpected that there is no ordering between logs from different containers (they just stream together in whatever order the lines are fetched)?

liggitt commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -293,13 +289,32 @@ func (o LogsOptions) RunLogs() error {
+				pipew.CloseWithError(err)
+			}
+
+			doneChan <- true
```
a waitgroup would be a cleaner way to do this

liggitt commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -293,13 +289,32 @@ func (o LogsOptions) RunLogs() error {
-		if err := o.ConsumeRequestFn(request, o.Out); err != nil {
-			return err
-		}
+		go func(request *rest.Request) {
```
is there a maximum number of parallel requests we should allow?

soltysh replied:
We need to have a default limit, with an option to change it. Otherwise by mistake we can dos the api server and that's something we don't want to happen.

liggitt commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -293,13 +289,32 @@ func (o LogsOptions) RunLogs() error {
-		}
+		go func(request *rest.Request) {
+			if err := o.ConsumeRequestFn(request, pipew); err != nil {
+				pipew.CloseWithError(err)
```
this closes all streams on first error. is that what we want?

m1kola reviewed (commented):
@liggitt, thanks for the review. I've updated the PR with changes (see the fixup commits since your review) and replied to your comments.

Please take another look.

m1kola commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -309,6 +320,10 @@ func DefaultConsumeRequest(request *rest.Request, out io.Writer) error {
-	return err
+	scanner := bufio.NewScanner(readCloser)
+	for scanner.Scan() {
+		fmt.Fprintf(out, "%s\n", scanner.Text())
```
Ah, e2e tests are failing with `--limit-bytes`, because of `\n` at the end. I'll have a look later.

m1kola commented:
@liggitt tests are passing now. Could you, please, take another look after the weekend?

Changes since your previous review: ba7cb09bcc36c1eb6acf398e9c65d4afefd1df40...e5549129c0db54659711e71317b3b3ed480ae65d

In these commits:

* A channel was replaced with a `WaitGroup` which helps to close `PipeWriter` once all log sources are exhausted
* `DefaultConsumeRequest` was modified to make sure that it doesn't interleave sub-line when running concurrently
* `DefaultConsumeRequest` was covered with unit tests

I'll squash commits before merging this

liggitt reviewed (commented):
one comment on error detection/handling on failed writes.

I'm still on the fence whether we want to support fan-out of long-running requests like this. @kubernetes/sig-cli-pr-reviews  @kubernetes/sig-scalability-pr-reviews, any thoughts on that?

m1kola commented on pkg/kubectl/cmd/logs.go:
```diff
@@ -293,22 +291,53 @@ func (o LogsOptions) RunLogs() error {
+	r := bufio.NewReader(readCloser)
+	for {
+		bytes, err := r.ReadBytes('\n')
+		out.Write(bytes)
```
Oops, sorry! I was going to return an error from the function to follow the previous behaviour as close as possible. I'll update tomorrow or on the weekend.

m1kola replied:
I fixed it in 1e830ad. Sorry for the delay.

m1kola commented:
/retest

m1kola commented:
Hi @liggitt, is there anything I can do to move this forward? I'm open for further feedback and ready to discuss alternative approaches to implement this.

liggitt commented:
I'd recommend raising the question with the CLI and scalability teams (either in slack, or on their mailing lists, or in one of their sig meetings) and see what feedback you get

pwittrock commented:
@m1kola please put this on the agenda for the next [sig-cli meeting](https://github.com/kubernetes/community/tree/master/sig-cli) (1 week from Wednesday.)

soltysh reviewed (commented):
Let's discuss in depth this on the next sig meeting on Oct 10th.

m1kola commented:
Hi. Yesterday, after the sig-cli meeting I:

* e77d1c1556aff5d27beae7b6f06cd8bbc156f58b - rebased the PR on top of the `master` branch
* 2255ba878e3c3f414ca0a323c2428d7a4e6de1c7 - added the flag that allows to control the number of concurrent of log streams with a (low) default limit

We also discussed the need to add the `--prefix` flag that will add source name to the log output. Everyone seemed to agree to handle this in a separate PR (I'll look into it once we are happy with this one).

m1kola commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -297,23 +301,67 @@ func (o LogsOptions) RunLogs() error {
-		if err := o.ConsumeRequestFn(request, o.Out); err != nil {
-			return err
-		}
+		go func(request *rest.Request) {
```
Hm. We should, probably read requests sequentially when `o.Follow` is false.

m1kola commented:
/retest

m1kola commented:
@soltysh and @seans3 I think, I've addressed the feedback you gave during the meeting. Could you, please, take another look at this PR and let me know, if you have any further comments.

I'm now looking into prefixing log lines with source (pod & container name) as we discussed, but it will be a separate PR to keep things simpler for reviewers and for me.

/assign @soltysh 
/assign @seans3

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -151,6 +158,7 @@ func NewCmdLogs(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 	cmd.Flags().StringVarP(&o.Container, "container", "c", o.Container, "Print the logs of this container")
 	cmdutil.AddPodRunningTimeoutFlag(cmd, defaultPodLogsTimeout)
 	cmd.Flags().StringVarP(&o.Selector, "selector", "l", o.Selector, "Selector (label query) to filter on.")
+	cmd.Flags().IntVar(&o.MaxFollowConcurency, "max-follow-concurency", o.MaxFollowConcurency, "Specify maximum number of concurrent logs to follow when using by a selector. Defaults to 5.")
```
I'd propose naming this `max-requests`

soltysh replied:
Maximum number of parallel log requests when using selector. Defaults to 5.

juanvallejo replied:
We could settle for a name in the middle, maybe something like `max-log-requests`

soltysh replied:
sgtm

m1kola replied:
@soltysh @juanvallejo I'm not sure that about `max-log-requests`: sound  too vague to me. When you do something like `kubectl logs -l deployment/some-name`, it will basically create N requests, but kubectl will be reading them sequentially.

Probably it's ok for end-users, because they don't care about the number of sequentially requests, right?

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -298,6 +302,47 @@ func (o LogsOptions) RunLogs() error {
+	if o.Follow && len(requests) > 1 {
+		if len(requests) > o.MaxFollowConcurency {
+			return fmt.Errorf(
+				"you are attempting to follow %d log streams, but maximum allowed concurency is %d. Use --max-follow-concurency to increase the limit",
```
Update when renaming the flag and use comma instead of a dot in that sentence.

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -298,6 +302,47 @@ func (o LogsOptions) RunLogs() error {
+}
+
+func (o LogsOptions) concurrentConsumeRequest(requests []rest.ResponseWrapper) error {
+	piper, pipew := io.Pipe()
```
`reader` and `writer` are better variable names

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -298,6 +302,47 @@ func (o LogsOptions) RunLogs() error {
+	return o.sequentialConsumeRequest(requests)
+}
+
+func (o LogsOptions) concurrentConsumeRequest(requests []rest.ResponseWrapper) error {
```
`parallelConsumeRequest`

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -298,6 +302,47 @@ func (o LogsOptions) RunLogs() error {
+		}(request)
+	}
+
+	go func() {
```
I think you want to block the main flow until you hear back from the wait group, iow. this code should not be part of any goroutine.

m1kola replied:
`io.Copy` below blocks the main flow. If we do not close pipe writer (`pipew` currently), it will be blocking the main flow forever (even if the server closed the connection). So we are waiting for all requests to finish in a separate goroutine and close the writer (as a result `io.Copy` stops blocking the main flow).

Or do I miss something?

soltysh commented on pkg/kubectl/cmd/logs/logs.go:
```diff
@@ -307,13 +352,33 @@ func (o LogsOptions) RunLogs() error {
+			return err
+		}
+
+		if err != nil {
```
```
if err != nil && err != io.EOF {
    return err
}
return nil
```

m1kola replied:
@soltysh `return nil` will stop the `for` loop after the first iteration. We want to return from the function when the first error appears, but we don't treat `io.EOF` as an error (because it is what we are waiting for). In case of no error we want to continue the loop.

soltysh reviewed (changes requested):
I left you some comments

juanvallejo commented:
@soltysh @m1kola how about naming the flag something like `max-log-requests`? At least to be a bit more descriptive about its intention

m1kola commented:
/test pull-kubernetes-integration

m1kola commented:
@soltysh please, take another look. I addressed your feedback in the latest changes and in comments.

And thanks for the review!

soltysh reviewed (approved):
/lgtm
/approve
@m1kola thanks for your patience :+1:

m1kola commented:
Test failures seem to be unrelated so let's try to rerun them.
/retest
